### PR TITLE
Fenix support

### LINKED ISSUE #624: Add support for downloading Firefox for Android (Fenix)
Given that the Fennec project is no longer available for quite some Firefox releases and got replaced with Fenix (GeckoView) we have to add support for downloading Fenix builds from archive.mozilla.org:

https://archive.mozilla.org/pub/fenix/

### PR DESCRIPTION
**Pull request by salilmishra23**
resolves #624

### CODE REVIEW

**Comment by salilmishra23** [comment]
Hey @whimboo! Can you please take a look and suggest changes if any, thanks!

**Comment by whimboo** [comment]
@salilmishra23 I'm absolutely sorry. When you created the PR I was out and then too many notifications have been received that I'm still working through the list. Sadly it's not that quick.

Now I have seen your PR and will make sure to review the code additions in a timely manner. I hope that you have still interest to continue in getting this features added to mozdownload!

File: mozdownload/scraper.py
Comment (whimboo) [inline mozdownload/scraper.py:422]: Please note that this would cause a failure for Fenix because for this application we do not have a `latest` folder. Instead we would have to retrieve the folder for the he latest date by scraping the subfolders with the latest year and then month. Once found another parsing step is needed to retrieve the list of builds from which we then can extract the date.

It means that the code here needs to set the url to eg. `nightly/2023/01/` and then get the listing for the given platform and pick the latest entry for the date regex, eg. `2022-12-31-17-02-16-fenix-110.0b1-android-x86_64/`. This should extract `2022-12-31` as date. See similar code in `get_build_info()`. Then we also have to return early with the found date to not trigger an exception in line 433

File: mozdownload/scraper.py
Comment (whimboo) [inline mozdownload/scraper.py:485]: Would you mind to create another global mapping for these two regex'es? That would allow us to not have different code paths here.

Otherwise downloading a Fenix build seem to work fine with the other issues from above applied and the binary run with `-t daily -a fenix -p android-x86_64 --date 2022-12-31`. \o/

File: mozdownload/scraper.py
Comment (whimboo) [inline mozdownload/scraper.py:558]: Please keep the timestamp in the filename that gets returned. Feel free to add another global mapper here as well for the formatting string.

File: tests/daily_scraper/test_daily_scraper.py
Comment (whimboo) [inline tests/daily_scraper/test_daily_scraper.py:137]: This seems to be a duplicated test. Probably it was meant for `android-armeabi-v7a`?

Also please add tests for release builds so that these can be tested as well. There are also remote tests which I would happy about to see 2 entries for the different build types as well.

**Review by whimboo — CHANGES_REQUESTED** [review]
Thank you for getting started to work on this issue! It's great to see the progress as already made! I added a couple of inline comments, so please have a look at these. Let me know if something is unclear and I'm happy to help. I'm looking really forward to a final version that we can merge. Thanks again!

**Comment by whimboo** [comment]
Please note that due to recent changes you should rebase your branch against master. Otherwise tests in CI will fail. Thanks.

**Comment by salilmishra23** [comment]
Hey whimboo! I tried to address all the comments you had. Please let me know if you want any additional changes.

**Review by whimboo — CHANGES_REQUESTED** [review]
I finally had a look after coming back from absence and it's great to see that tests are passing. Sadly there are some further issues that need to be addressed given that right now I'm not able to download actual fenix builds from the remote server.

1) Please try to download various builds yourself. Like downloading a release build of Fenix via `mozdownload -a fenix -v latest -p android-x86_64`. It currently produces a wrong path so the build cannot be found under `https://archive.mozilla.org/pub/fenix/releases/110.1.0/android/fenix-110.1.0-android-x86_64/`. As such we would need real remote tests similar to other applications.

2) If Fennec is a problem I'm happy to just replace it's support with Fenix. Fennec is old and cannot be used anymore anyway. If someone needs it then an older version of mozdownload could be used.

Please let me know if you have questions.

**Comment by salilmishra23** [comment]
Hey @whimboo, it would be great if you can review the PR again, thanks!

File: mozdownload/scraper.py
Comment (whimboo) [inline mozdownload/scraper.py:36]: I still see a problem with downloading here when using `mozdownload -a fenix -v latest -p android-x86_64`. In this case we do not have a mapper from fenix to a version class. I had a look at mozilla-version but there is actually no specific class for Fenix/GeckoView. We cannot use FennecVersion because that's a different product. 

I'll sort that out with the maintainers, so could you for now please add a mapping from `fenix` to `FirefoxVersion`? Thanks.

File: mozdownload/scraper.py
Comment (whimboo) [inline mozdownload/scraper.py:65]: All those older APIs were only needed for Fennec and can be removed. The earliest [Fenix beta is 97.0.0](https://archive.mozilla.org/pub/fenix/releases/97.0.0-beta.1/android/) and we should support all its flavors only.

File: mozdownload/scraper.py
Comment (whimboo) [inline mozdownload/scraper.py:80]: Same here as well.

File: mozdownload/scraper.py
Comment (whimboo) [inline mozdownload/scraper.py:563]: Please remove all those as well.

File: mozdownload/scraper.py
Comment (whimboo) [inline mozdownload/scraper.py:693]:
```suggestion
        if self.application == "fenix":
            regex = r'releases/%(VERSION)s/android/fenix-%(VERSION)s-%(PLATFORM)s/'
        else:
            regex = r'releases/%(VERSION)s/%(PLATFORM)s/%(LOCALE)s/'
```

File: tests/remote/manifest.ini
Comment (whimboo) [inline tests/remote/manifest.ini:1]: Please not just remove this test but instead rename it so that we can have remote tests to ensure that real downloads pass as well.

**Review by whimboo — CHANGES_REQUESTED** [review]
This is a great update and I think by removing Fennec it made it much more simpler! Thanks for that work!

Please have a look at the inline comments that I've made. It's not major just some clean-ups and small fixes.

Beside that you should still have a look at the remote tests for Fenix. Those we definitely need to ensure our product is working as example:

* Please check the tests folder for left-over `mobile` traces, eg. remove the `tests/data/mobile` folder
* Fix the path for `mobile` to `fenix` in `test_api.py`.
* Also a plus point if you could add unit tests for the release scraper for Fenix. Test files should be placed in `tests/data/fenix/releases`.

Once that's done there is nothing more left and the PR can be merged. Thanks again!

File: mozdownload/scraper.py
Comment (whimboo) [reply mozdownload/scraper.py:36]: I've created https://github.com/mozilla-releng/mozilla-version/pull/105 to more easily can import `MobileVersion`. Unless that has been merged and released, lets use `FirefoxVersion` as mentioned above.

Comment (whimboo) [reply mozdownload/scraper.py:36]: This PR has been merged now. But we have to wait for the next release. Given that I don't know when this will happen please use `FirefoxVersion` as proposed and I'll change it later on.

**Comment by whimboo** [comment]
Hi @salilmishra23. I wanted to ask if you may have the time to get the remaining review comments addressed. I would like to release a new version of mozdownload soon and it would be great to have this feature included. Thanks!

**Comment by salilmishra23** [comment]
Hey @whimboo, sorry for the long delay. I have addressed all the comments, please have a look!

File: mozdownload/scraper.py
Comment (salilmishra23) [reply mozdownload/scraper.py:36]: added the fix

**Comment by whimboo** [comment]
> Hey @whimboo, sorry for the long delay. I have addressed all the comments, please have a look!

Hi @salilmishra23 no worries. Great that you found the time to continue on this PR! I appreciate. Now that I'm back from days off due to holidays as well let me take a look.

File: mozdownload/scraper.py
Comment (whimboo) [reply mozdownload/scraper.py:36]: This looks great now. Thanks!

File: tests/remote/test_fenix.py
Comment (whimboo) [inline tests/remote/test_fenix.py:20]: The tests here are not complete because it misses the tests for release and beta builds similar to Firefox and Thunderbird.

**Review by whimboo — COMMENTED** [review]
Overall this looks great now! All the variations of types, platforms work fine for Fenix. Thanks a lot!

There is just a small inline comment regarding missing tests. Let me know if you would be able to add these. Otherwise I can do myself after merging this PR.

**Comment by salilmishra23** [comment]
Thanks for the reviewing! At the moment, I can't find time to add the required tests, please feel free to do the needful.

**Comment by whimboo** [comment]
> Thanks for the reviewing! At the moment, I can't find time to add the required tests, please feel free to do the needful.

Sorry for the delay. I think it's fine then when we move the creation of the additional tests to a follow-up issue / PR. I've filed https://github.com/mozilla/mozdownload/issues/663 for it and I will take care of it.

Thanks a lot for your contribution and the time you spent on learning about mozdownload. With the next release it will definitely help a lot of users in getting those builds!

**Review by whimboo — APPROVED** [review]

**Pull request merged**
